apps/dispatch/dispatch-1.py

```diff
@@ -0,0 +1,275 @@
+
+			import GafferUI
+			import GafferDispatchUI
+
+			self.__dialogue = GafferDispatchUI.DispatchDialogue( script, nodes, dispatcherType, applyUserDefaults=args["applyUserDefaults"].value )
+			self.__dialogueClosedConnection = self.__dialogue.closedSignal().connect( Gaffer.WeakMethod( self.__dialogueClosed ) )
+			dispatcher = self.__dialogue.getDispatcher()
+
+		else :
+
+			dispatcher = GafferDispatch.Dispatcher.create( dispatcherType )
+			if not dispatcher :
+				IECore.msg( IECore.Msg.Level.Error, "gaffer dispatch", "{} is not a registered dispatcher.".format( dispatcherType ) )
+				return 1
+
+			if args["applyUserDefaults"].value :
+				Gaffer.NodeAlgo.applyUserDefaults( dispatcher )
+
+		if len(args["settings"]) % 2 :
+			IECore.msg( IECore.Msg.Level.Error, "gaffer dispatch", "\"settings\" parameter must have matching entry/value pairs" )
+			return 1
+
+		for i in range( 0, len(args["settings"]), 2 ) :
+			key = args["settings"][i].lstrip( "-" )
+			value = args["settings"][i+1]
+			if key.startswith( "context." ) :
+				entry = key.partition( "context." )[-1]
+				status = self.__setValue( entry, value, script, context=True )
+			elif key.startswith( "dispatcher." ) :
+				identifier = key.partition( "dispatcher." )[-1]
+				status = self.__setValue( identifier, value, dispatcher )
+			else :
+				status = self.__setValue( key, value, script )
+			if status :
+					return status
+
+		if args["gui"].value :
+
+			self.__dialogue.setVisible( True )
+
+			GafferUI.EventLoop.mainEventLoop().start()
+
+		else :
+
+			return self.__dispatch( dispatcher, nodes )
+
+		return 0
+
+	@staticmethod
+	def __dispatch( dispatcher, nodes ) :
+
+		script = nodes[0].scriptNode()
+
+		try :
+
+			with script.context() :
+				dispatcher.dispatch( nodes )
+
+		except Exception :
+
+			IECore.msg(
+				IECore.Msg.Level.Error,
+				"gaffer dispatch : dispatching %s" % str( [ node.relativeName( script ) for node in nodes ] ),
+				"".join( traceback.format_exception( *sys.exc_info() ) ),
+			)
+
+			return 1
+
+		return 0
+
+	def __dialogueClosed( self, dialogue ) :
+
+		import GafferUI
+
+		GafferUI.EventLoop.mainEventLoop().stop()
+
+	@staticmethod
+	def __setValue( identifier, value, parent, context=False ) :
+
+		if context :
+			parent["variables"].addMember( identifier, eval( value ) )
+			return 0
+
+		plug = parent.descendant( identifier )
+		if not plug :
+			IECore.msg( IECore.Msg.Level.Error, "gaffer dispatch", "\"%s\" does not contain a plug named \"%s\"." % ( parent.getName(), identifier ) )
+			return 1
+		if not plug.settable() :
+			IECore.msg( IECore.Msg.Level.Error, "gaffer dispatch", "\"%s\" cannot be set." % identifier )
+			return 1
+
+		try :
+			plug.setValue( eval( value ) )
+		except Exception as exception :
+			IECore.msg( IECore.Msg.Level.Error, "gaffer dispatch : setting \"%s\"" % identifier, str( exception ) )
+			return 1
+
+		return 0
+
+	@staticmethod
+	def __create( path ) :
+
+		path = path.split( "." )
```

Comment:
Addressed in 8c1312abadc2b0dd453ce4921754e03a8dbcd524. Note I didn't add that `GafferDispatch` convenience, because I don't want to make it easy to forget that its inconvenient for all IE nodes.

Reply:
This is still giving nasty errors for some values :

```
> gaffer dispatch -nodes GafferDispatch

gaffer Traceback (most recent call last):
  File "/disk1/john/dev/build/gaffer/bin/gaffer.py", line 169, in <module>
    result = app.run()
  File "/disk1/john/dev/build/gaffer/python/Gaffer/Application.py", line 110, in run
    return self.__run()
  File "/disk1/john/dev/build/gaffer/python/Gaffer/Application.py", line 142, in __run
    return self._run( self.parameters().getValidatedValue() )
  File "/disk1/john/dev/build/gaffer/apps/dispatch/dispatch-1.py", line 164, in _run
    node = self.__create( nodeName )
  File "/disk1/john/dev/build/gaffer/apps/dispatch/dispatch-1.py", line 294, in __create
    return result()
TypeError: 'module' object is not callable
```

```
> gaffer dispatch -nodes GafferDispatch.Oops

Traceback (most recent call last):
  File "/disk1/john/dev/build/gaffer/bin/gaffer.py", line 169, in <module>
    result = app.run()
  File "/disk1/john/dev/build/gaffer/python/Gaffer/Application.py", line 110, in run
    return self.__run()
  File "/disk1/john/dev/build/gaffer/python/Gaffer/Application.py", line 142, in __run
    return self._run( self.parameters().getValidatedValue() )
  File "/disk1/john/dev/build/gaffer/apps/dispatch/dispatch-1.py", line 164, in _run
    node = self.__create( nodeName )
  File "/disk1/john/dev/build/gaffer/apps/dispatch/dispatch-1.py", line 292, in __create
    result = getattr( result, n )
AttributeError: 'module' object has no attribute 'Oops'
```